Bug fixes regarding schedule storage implementation

kevin-wijnen opened the pull request:
Fix: Duplicate schedules for a single schedule of a relay
Fix: TimeOff and TimeOffType properly returned with schedule.toString()
Fix: Schedule deletion no more needing a second data message

kevin-wijnen committed:
Bug fixes regarding schedule storage implementation

Fix: Duplicate schedules for a single schedule of a relay
Fix: TimeOff and TimeOffType properly returned with schedule.toString()
Fix: Schedule deletion no more needing a second data message

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/RelaySchedulePair.java:
```diff
@@ -0,0 +1,58 @@
+        return this.relayNr;
+    }
+
+    public void setRelayNr(int relayNr) {
+        this.relayNr = relayNr;
+    }
```
setRelayNr and the setScheduleNr should be removed. After initialising a RelaySchedulePair, it should not be possible to modify it.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/RelaySchedulePair.java:
```diff
@@ -0,0 +1,58 @@
+package com.cgi.iec61850serversimulator.dataclass;
+
+public class RelaySchedulePair {
+    /**
+     * Class to store relay numbers and schedule numbers into a single object for
+     * EventDataListener HashMap and general refactoring of code
+     */
```
This JavaDoc should be right above the "public class" line. Generate the JavaDoc for your application locally and check what comes out. The JavaDoc here is probably added as explanation for the relayNr field.

kevin-wijnen replied:
Done, however this JavaDoc was manually created as an explanation for what purpose this class serves.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -234,17 +234,20 @@ public void updateDatabaseSchedule(Schedule schedule) {
         int relayNr = schedule.getRelayNr();
-        boolean found = this.scheduleRepository.existsByIndexNumberAndRelayId(schedule.getIndexNumber(), relayNr);
+        List<RelayEntity> relayEntities = this.relayRepository.findByIndexNumber(relayNr);
+        RelayEntity relayEntity = relayEntities.get(0);
+        long relayId = relayEntity.getId();
+        boolean found = this.scheduleRepository.existsByIndexNumberAndRelayId(schedule.getIndexNumber(), relayId);
 
         if (found) {
             List<ScheduleEntity> scheduleEntities = this.scheduleRepository.findByIndexNumberAndRelayId(indexNumber,
-                    relayNr);
+                    relayId);
```
Haven't you got this code (or almost exactly the same) somewhere else? If so, turn it into a function.

kevin-wijnen replied:
Done by having findRelayIdByRelayNr to not have to initialise the RelayEntity list every time to obtain its relayId.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -48,6 +52,7 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
     public List<ServiceError> write(final List<BasicDataAttribute> bdas) {
         logger.info("BDA write request scanning...");
         boolean modified = false;
+        HashMap<RelaySchedulePair, Schedule> toUpdateSchedules = new HashMap<RelaySchedulePair, Schedule>();
```
Change this to:
Map<RelaySchedulePair, Schedule> toUpdateSchedules = new HashMap<>();

1) Declaring a variable as an interface (like Map) makes it slightly easier to maintain.
2) Just using "<>" when constructing the HashMap saves some typing.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -66,6 +71,8 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
                     scheduleNr = this.extractScheduleIndex(bda.getReference());
                 }
 
+                RelaySchedulePair relaySchedulePairs = new RelaySchedulePair(relayNr, scheduleNr);
```
relaySchedulePairs => relaySchedulePair, because it contains exactly one pair.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -48,6 +52,7 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
     public List<ServiceError> write(final List<BasicDataAttribute> bdas) {
         logger.info("BDA write request scanning...");
         boolean modified = false;
+        HashMap<RelaySchedulePair, Schedule> toUpdateSchedules = new HashMap<RelaySchedulePair, Schedule>();
```
"schedulesToUpdate" sounds more natural than "toUpdateSchedules".

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -173,9 +180,9 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
-                                .updateDatabaseSchedule(this.device.getRelay(relayNr).getSchedule(scheduleNr));
+                        // Puts to update schedule in HashMap
+                        toUpdateSchedules.put(relaySchedulePairs,
+                                this.device.getRelay(relayNr).getSchedule(scheduleNr));
```
If you add something like this in the beginning of the for-loop, you can simplify lots of "this.device.getRelay(relayNr).getSchedule(scheduleNr)" calls to just "scheduleToUpdate":
Schedule scheduleToUpdate;
"if (relayNr > 0 && scheduleNr > 0) {
  scheduleToUpdate = this.device.getRelay(relayNr).getSchedule(scheduleNr);
}"

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -329,6 +331,16 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
             }
+            // Iterate through toUpdateSchedules to update the schedules
+            for (Entry<RelaySchedulePair, Schedule> entry : toUpdateSchedules.entrySet()) {
+                RelaySchedulePair key = entry.getKey();
+                int relayNr = key.getRelayNr();
+                Schedule value = entry.getValue();
+                int scheduleNr = value.getIndexNumber();
+
+                this.databaseUtils.updateDatabaseSchedule(this.device.getRelay(relayNr).getSchedule(scheduleNr));
```
This can be simplified. The entries in your Map are exactly the Schedule objects you need to update in the database.

kevin-wijnen replied:
Done

kevin-wijnen committed:
Pull Request feedback integrated

Layout: Moved RelaySchedulePair JavaDoc above the class
Layout: Changed relaySchedulePairs to relaySchedulePair
Layout: Changed toUpdateSchedule to scheduleToUpdate (same for plural)

Refactor: Removed setters from RelaySchedulePair
Refactor: Turned getting relayId from relayNr into function to remove
duplicate code
Refactor: Changed toUpdateSchedules initialization to simplify
Refactor: Included if-check to have scheduleToUpdate ready at beginning
of for-loop in EventDataListener
Refactor: Simplification of updating schedules with RelaySchedulePair

kevin-wijnen committed:
Name of function correction

Layout: Changed findRelayEntityByRelayNr to findRelayIdByRelayNr

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -220,13 +231,11 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
                         logger.info("Time On Type found.");
-                        this.device.getRelay(relayNr)
-                                .getSchedule(scheduleNr)
+                        this.device.getRelay(relayNr).getSchedule(scheduleNr)
```
You can use the schedulesToUpdate variable here as well. You can use it in all places, where you have "this.device.getRelay(relayNr).getSchedule(scheduleNr)".

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -329,6 +330,12 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
                     }
                 }
             }
+            // Iterate through toUpdateSchedules to update the schedules
```
The comment doesn't really add anything, remove it.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -329,6 +330,12 @@ public EventDataListener(final Device device, final Scheduler scheduler, final D
                 }
             }
+            // Iterate through toUpdateSchedules to update the schedules
+            for (Entry<RelaySchedulePair, Schedule> relaySchedulePair : schedulesToUpdate.entrySet()) {
```
Even simpler:
for (Schedule scheduleToUpdate : schedulesToUpdate.values()) {
  this.databaseUtils.updateDatabaseSchedule(scheduleToUpdate);
}

kevin-wijnen replied:
Done

kevin-wijnen committed:
Pull Request feedback processed

Refactor: Using schedulesToUpdate variable more to simplify code
Refactor: Simplified for-loop for updating schedules in database

Layout: Removed unnecessary comment about schedule updating
